policyeval/protections.go

```diff
@@ -0,0 +1,749 @@
+			// Registration is disabled or handled externally
+			pe.Bot.Log.Trace().Stringer("user_id", target).Msg("homeserver forbids registration.")
+			return false, nil
+		}
+		pe.Bot.Log.Err(err).Str("homeserver", hs).Msg("failed to query registration flows for insecure registration check")
+		return false, err
+	}
+
+	hit = flows.HasSingleStageFlow(mautrix.AuthTypeDummy)
+	pe.Bot.Log.Trace().
+		Str("homeserver", hs).
+		Any("flows", flows.Flows).
+		Bool("dangerous", hit).
+		Msg("server has registration enabled")
+	if hit {
+		// Kick user and alert the management room
+		if !dry {
+			go i.Kick(ctx, pe, evt, target)
+		} else {
+			pe.sendNotice(
+				ctx,
+				fmt.Sprintf(
+					"[DRY RUN] Would have kicked [%s](%s) from [%s](%s) for joining from a homeserver (%s) that allows insecure registration.",
+					target,
+					target.URI(),
+					evt.RoomID,
+					evt.RoomID.URI(),
+					target.Homeserver(),
+				),
+			)
+		}
+	}
+	return hit, nil
+}
+
+// AntiFlood is a protection that redacts and bans users who send too many events in a given time period.
+type AntiFlood struct {
+	Limit          int              `json:"limit"` // how many events to allow before actioning
+	Per            jsontime.Seconds `json:"per"`   // the timespan in which to count events
+	MaxInfractions int              `json:"max_infractions,omitempty"`
+	TrustServer    bool             `json:"trust_server,omitempty"` // if false, use local time, instead of evt origin
+	counts         map[id.UserID]int
+	expire         map[id.UserID]time.Time
+	countLock      sync.Mutex
+}
+
+func (a *AntiFlood) Execute(ctx context.Context, pe *PolicyEvaluator, evt *event.Event, dry bool) (hit bool, err error) {
+	if a.Limit <= 0 || evt.StateKey != nil {
+		return false, nil // no-op
+	}
+	a.countLock.Lock()
+	defer a.countLock.Unlock()
+	if a.counts == nil {
+		a.counts = make(map[id.UserID]int)
+	}
+	if a.expire == nil {
+		a.expire = make(map[id.UserID]time.Time)
+	}
+
+	// Expire old counts
+	now := time.Now()
+	origin := time.UnixMilli(evt.Timestamp)
+	if !a.TrustServer {
+		origin = now
+	}
+	for user, exp := range a.expire {
+		if now.After(exp) {
+			delete(a.counts, user)
+			delete(a.expire, user)
+		}
+	}
+
+	// Count event
+	a.counts[evt.Sender]++
+	expire, ok := a.expire[evt.Sender]
+	if !ok || expire.Before(origin) {
+		// If there isn't already an expiry, or the current expiry is before the event origin, set a new expiry
+		expire = time.UnixMilli(evt.Timestamp).Add(a.Per.Duration)
+	}
+	a.expire[evt.Sender] = expire
+	pe.Bot.Log.Trace().
+		Str("protection", "anti_flood").
+		Stringer("sender", evt.Sender).
+		Stringer("room_id", evt.RoomID).
+		Stringer("event_id", evt.ID).
+		Int("count", a.counts[evt.Sender]).
+		Time("expires", expire).
+		Int("infractions", a.counts[evt.Sender]-a.Limit).
+		Msg("anti_flood count and expiry updated")
+
+	if a.counts[evt.Sender] > a.Limit {
+		hit = true
+		infractions := a.counts[evt.Sender] - a.Limit
+		if !dry {
```

Comment:
I recommend splitting the handles on each protection and standardising the dispatch mechanism for the dry and wet handles in all protections. Because this check is adhoc and for now it looks simple but as the code base ages and moves, and new protections or behaviours are added, the risk that wet code ends up in a dry handle will only increase